README.md
# RestVolley
-------

A http request engine based on [Volley] and [OkHttp], giving up Apache HttpClient request. supports image loading, restful api requesting, and file downloading. [READ MORE]

# Dependency
-------

binary on [jcenter](http://jcenter.bintray.com/com/hujiang/restvolley/restvolley/)

dependency with Gradle:

```
compile 'com.hujiang.restvolley:restvolley:1.0.11'
```

# Functions
-------
## Restful API request
RestVolley make restful API requesting conveniently.It support the most http method request, such as `GET`,`HEAD`,`POST`,`DELETE`,`OPTIONS`,`PATCH`,`PUT`,`TRACE`.

|| GET | POST | DELETE | PUT | HEAD | PATCH | OPTIONS | TRACE |
|:---:| :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
|Cacheable|Y|N|N|N|Y|N|Y|Y|
|requireBody|N|Y|N|Y|N|Y|N|N|
|permitBody|||Y||||||

* create a request

```
//Get request
new GetRequest(context).url("").execute(String.class, new RestVolleyCallback<String>());
//post request
new PostRequest(context).url("").execute(String.class, new RestVolleyCallback<String>());
//Delete request
new DeleteRequest(context).url("").execute(String.class, new RestVolleyCallback<String>());
//Put request
new PutRequest(context).url("").execute(String.class, new RestVolleyCallback<String>());
//Head request
new HeadRequest(context).url("").execute(String.class, new RestVolleyCallback<String>());
//Patch request
new PatchRequest(context).url("").execute(String.class, new RestVolleyCallback<String>());
//Options request
new OptionsRequest(context).url("").execute(String.class, new RestVolleyCallback<String>());
//Trace request
new TraceRequest(context).url("").execute(String.class, new RestVolleyCallback<String>());
```
* set http request params: header, urlparams, requestbody, contentType, charset

```
new GetRequest(context).url("")
	.setTag("")
	.setRetryPolicy(new DefaultRetryPolicy())
	.setShouldCache(true)
	.setPriority(Priority.Normal)
	//default is "application/json"
	.setContentType("application/json")
	//default is "UTF-8"
	.setCharset("UTF-8")
	.addHeader("userId", "12345")
	.addHeaders(map)
	.setUserAgent("")
	.addParams(key, value)
	.setCacheEntry(...)
	.execute(.....);
```
* Cache

```
//default false
setShouldCache(true);
setCacheEntry(...)
```
* RetryPolicy

```
setRetryPolicy(new DefaultRetryPolicy())
```
* Priority

```
//default is Priority.Normal
setPriority(Priority.Normal)
```
* ContentType and Charset

```
	//default is "application/json"
	setContentType("application/json")
	//default is "UTF-8"
	setCharset("UTF-8")

```

* Tag

```
setTag(tag);
```

* UserAgent

```
setUserAgent("")
```
* timeout

```
setConnectTimeout(...);
setReadTimeout(...);
setWriteTimeout(...);

//for all timout
setTimeout(...);

```
* header

```
addHeader(key, value);
//or
addHeaders(map);
```
* params

```
//add url params
addParams(key, value);

//for post/put/patch or delete
addParams(key, file);
addParams(key, inputstream);
```
* body and HttpEntity

```
//body and HttpEntity for post/put/patch or delete
setBody(byte[] bytes);
setBody(HttpEntity httpEntity);
setBody(String s);

//convert params to form data entity.
paramsToFormEntity();
//convert params to json entity.
paramsToJsonEntity();
//convert params to multipart entity.
paramsToMultipartEntity();

```

* execute and syncExecute

```
//async execute with callback.
new GetRequest(context).url().execute(String.class, new RestVolleyCallback<String>());

//sync execute.
RestVolleyResponse<String> response = new GetRequest(context).url().syncExecute(String.class);
```
* request with new RequestEngine

suggest using the default RequestEngine, only if the request need proxy.

```
//
new GetRequest(context).setRequestEngine(...);

//create a new RequestEngine
RestVolley.newRequestEngine(context, engineTag);
```
## Image load
RestVolleyImageLoader has memory cache and disk cache, supports not only network image request, but sdcard/assets/res/ContentProvider Uri image request, uri scheme like:

|||
|:---|:---|
|http/https|http://www.google.com/xxx/xxx.png|
|file|file:///mnt/sdcard/xxx.png|
|assets|assets://assets_default.png|
|res|“drawable://” + R.drawable.drawable_default, or "drawable://" + R.raw.defaut|
|ContentProvider|content://media/external/images/media/27916|

* set global config

```
RestVolleyImageLoader.instance(context).setConfig(ImageLoaderGlobalConfig cfg);
```
* ImageLoaderGlobalConfig

```
ImageLoaderGlobalConfig.create().memCacheSize(..).diskCacheSize(..).diskCacheDir(..).requestEngine(..);
```
* ImageLoadOption

```
ImageLoadOption.create().defaultImgResId(..)
	.errorImgResId(..)
	.imgLoadAnimation(..)
	.scaleType(..)
	.maxWidth(..)
	.maxHeight(..)
	.cacheEnable(..);
```

* displayImage

```
RestVolleyImageLoader.instance(context).displayImage(uri, imageView);
//
RestVolleyImageLoader.instance(context).displayImage(uri, imageView, imageLoadOption);
```

* loadImage

```
RestVolleyImageLoader.instance(context).loadImage(uri, ImageListener);
//
RestVolleyImageLoader.instance(context).loadImage(uri, ImageLoadOption, ImageListener);
```
* syncLoadImage

```
RestVolleyImageLoader.instance(context).syncLoadImage(String uri);
//
RestVolleyImageLoader.instance(context).syncLoadImage(uri, ImageLoadOption);
```
* others

```
RestVolleyImageLoader.instance(context).isCached(...);
RestVolleyImageLoader.instance(context).removeCache(...);

```

## File download

```
new RestVolleyDownload(context)
	.url(url)
	.addHeader(key, value)
	.tag(tag)
	.setProxy(...)
	.setTimeout(...)
	.setIsAppend(..)
	.download(......)
	//or
	.syncdownload(....)
```

simple static download.

```
RestVolleyDownload.download(....)
//or
RestVolleyDownload.syncDownload(....)
```

## Data upload
Using RestVolleyUpload as PostRequest.

```
new RestVolleyUpload(context).url(..).execute(...);
//or
new RestVolleyUpload(context).url(..).syncExecute(...);
```

# License
-------
```
Copyright (C) 2016 Hujiang, Inc.
Copyright (C) 2014 Xiaoke Zhang.
Copyright (C) 2011 The Android Open Source Project.
Copyright (C) 2016 Square, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```

# [Change Log](CHANGELOG.md)
------


























[Volley]:https://github.com/mcxiaoke/android-volley
[OkHttp]:https://github.com/square/okhttp
[READ MORE]:docs/OverView.md

CHANGELOG.md
### v1.2.2(2019-01-14)
* fix: 同一个url图片地址，无法收到回调的BUG

### v1.2.1(2018-10-24)
* add: uses-library "org.apache.http.legacy"
* add: imageloader添加只删除缓存的api
* fix: 图片下载未回调的BUG

### v1.1.11(2018-06-22)
* remove duplicated item `Content-Type` in request headers

### v1.1.10(2018-05-29)
* fix: NullPointException on LinkedList$Node.next

### v1.1.9(2018-04-02)
* [fix] StringIndexOfBoundException: length=0;index=-1

### v1.1.8(2018-04-02)
* [update]默认不信任签名证书

### v1.1.7(2018-01-08)
* [fix]解决将json字符串反序列化为map时integer被强制转成double的问题

### v1.1.6(2017-10-13)
* [fix] NullPointException occurs while adding null cookie to cookie store

### v1.1.5(2017-09-21)
* [fix] cookie null check

### v1.1.4(2017-09-19)
* public RestVolleyRequest.WEBAPI_REQUEST_ENGINE

### v1.1.3(2017-09-13)
* cookie supported

### v1.1.2(2017-08-30)
* fix protocol exception

### v1.1.1(2017-08-18)
* add call.setException when request fail

### v1.1.0(2017-08-14)
* [add] add callback to syncExecute methods

### v.1.0.22(2017-06-13)
* [add] add method setProxy to RestVolleyRequest

### v1.0.21(2017-05-08)
*[fix] NullPointException on RestVolleyImageCache.getDiskCacheDir(Context context, String uniqueName)

## v1.0.20(2017-04-28)
* [add]`DeleteRequest`支持`body`，使用构造函数`DeleteRequest(Context context, boolean isBodyEnable)`支持`body`

### v1.0.19(2017-03-09)
* 修复接口RestVolleyImageLoader.isCached(..)返回状态不对的Bug

### v1.0.18(2017-03-08)
* 修复gzip 压缩参数Content-Encoding, Accept-Encoding支持的Bug

### v1.0.17(2017-02-22)
* 添加api GsonUtils.fromJsonStringThrowEx， GsonUtils.optFromJsonString，GsonUtils.toJsonStringThrowEx，GsonUtils.optToJsonString
* 优化Demo
* 优化联网代理问题

### v1.0.16(2017-02-07)
* 解决部分4.+手机固件上webp图片加载的问题

### v1.0.15(2017-01-16)
* 解决NullPointerException的问题

### v1.0.14(2017-01-13)
* 修复无网络情况下同步接口出现NullException的Bug

### v1.0.13(2017-01-09)
* 修复1.0.12版本引起的超时重试时请求队列被阻塞的问题

### v1.0.12(2016-12-29)
* 优化图片内存问题
* 解决清除图片缓存时DiskLruCache出现异常：IllegalStateException:cache is closed
* RestVolleyImageLoader支持流式图片下载，降低内存占用率
* 优化https支持, 默认信任所有请求，安全要求较高的app可以通过接口设置SSLSocketFactory和HostnameVerifier，或者添加证书锁定CertificatePinner来增加安全性
* 提供常量CertificateUtils.DEFAULT_HOSTNAME_VERIFIER; CertificateUtils.ALLOW_ALL_HOSTNAME_VERIFIER
* 提供接口CertificateUtils.getDefaultSSLSocketFactory()
* 支持内置自签名证书CertificateUtils.getSSLSocketFactory(InputStream certificateInputStream)
* 优化下载逻辑，支持HEAD头不传Content-Length的下载

### v1.0.11(2016-12-28)
* 解决RestVolleyImageLoader加载本地大图出现OOM的问题，但是网络大图下载的问题暂时没有解决，后续会重写网络图片下载那块逻辑，解决大图下载问题

### v1.0.10(2016-12-21)
 * 解决当网络请求发生VolleyError时同步接口syncExecute(...)返回数据异常的Bug

### v1.0.9(2016-11-28)
* fix: no onDownloadFailure callback when SDCard is full
* recover API RestVolleyDownload#cancel(tag)

### v1.0.8(2016-11-07)
* add method RestVolleyImageLoader#getDiskCacheDir(), RestVolleyImageLoader#getDiskCachePath()
* add method RestVolleyImageLoader#displayImage(final String uri, final ImageView imageView, final ImageDisplayer imageDisplayer)
* add metnod RestVolleyImageLoader#displayImage(final String uri, final ImageView imageView, final ImageLoadOption option, final ImageDisplayer displayer)
* remove ImageDisplayer from ImageLoadOption
* RestVolleyDownload增加下载临时文件，后缀为.tmp
* 增加取消下载的接口RestVolleyDownload#cancel(), 删除接口RestVolleyDownload#cancel(Object tag)
* fix bug: java.lang.IllegalStateException: cache is closed on DiskLruCache.java:614
* add method RestVolleyImageLoader#getDiskCachePath(......)

### v1.0.7(2016-10-26)
* add ImageLoadOption#imageDisplayer, ImageLoadOption#bitmapConfig
* add method RestVolleyImageLoader#removeAllCache(), RestVolleyImageLoader#removeDiskCache()
* add method RestVolleyImageLoader#start(), RestVolleyImageLoader#stop()

### v1.0.6(2016-10-19)
* 解决磁盘IO异常导致磁盘缓存失效，并导致程序崩溃的问题

### v1.0.5
* 添加接口RestVolleyCallback#getException()获取异常信息

### v1.0.4
* 解决同步下载接口不能下载文件的问题

### v1.0.1
* fix bintrayUpload error

### v1.0.0
* release for rest api requesting, image loading, downloading, uploading.




app/src/main/java/com/hujiang/normandy/MainActivity.java
package com.hujiang.normandy;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;

import com.android.volley.VolleyError;
import com.hujiang.account.AccountManager;
import com.hujiang.account.api.AccountAPI;
import com.hujiang.account.api.AccountApiCallBack;
import com.hujiang.account.api.BaseAccountModel;
import com.hujiang.account.api.model.UserInfo;
import com.hujiang.account.app.LoginActivity;
import com.hujiang.account.app.MyAccountActivity;
import com.hujiang.account.html5.LoginJSEvent;
import com.hujiang.common.concurrent.TaskScheduler;
import com.hujiang.common.util.LogUtils;
import com.hujiang.common.util.ToastUtils;
import com.hujiang.normandy.api.DemoAPI;
import com.hujiang.restvolley.download.RestVolleyDownload;
import com.hujiang.restvolley.image.ImageLoaderCompat;
import com.hujiang.restvolley.image.RestVolleyImageLoader;
import com.hujiang.restvolley.webapi.RestVolleyCallback;
import com.hujiang.restvolley.webapi.RestVolleyResponse;
import com.squareup.okhttp.Headers;

import java.io.File;
import java.util.Map;

//circleid 136651365l
public class MainActivity extends AppCompatActivity implements AccountManager.AccountObserver {

    private Button mAccountButton;
    private Button mLeagueButton;
    private Button mFileImgButton;
    private Button mAssetsImgButton;
    private Button mContentImgButton;
    private Button mDrawableImgButton;
    private Button mRawImgButton;
    private Button mDownloadButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        AccountManager.instance().registerAccountObserver(this);
        LoginJSEvent.getInstance().registerContext(this);

        mAccountButton = (Button)findViewById(R.id.account_btn);

        mLeagueButton = (Button)findViewById(R.id.league_btn);
        mFileImgButton = (Button)findViewById(R.id.file_img_btn);
        mAssetsImgButton = (Button)findViewById(R.id.asset_img_btn);
        mContentImgButton = (Button)findViewById(R.id.content_img_btn);
        mDrawableImgButton = (Button)findViewById(R.id.drawable_btn);
        mRawImgButton = (Button)findViewById(R.id.raw_btn);
        mDownloadButton = (Button)findViewById(R.id.download_btn);

        mLeagueButton.setOnClickListener(mOnClickListener);
        mFileImgButton.setOnClickListener(mOnClickListener);
        mAssetsImgButton.setOnClickListener(mOnClickListener);
        mContentImgButton.setOnClickListener(mOnClickListener);
        mDrawableImgButton.setOnClickListener(mOnClickListener);
        mRawImgButton.setOnClickListener(mOnClickListener);
        mDownloadButton.setOnClickListener(mOnClickListener);

        mAccountButton.setText(AccountManager.instance().isLogin() ? "个人中心" : "登录");
        mAccountButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (AccountManager.instance().isLogin()) {
                    MyAccountActivity.start(MainActivity.this, true, true, true, true);
                } else {
                    LoginActivity.start(MainActivity.this);
                }
            }
        });

        findViewById(R.id.get_btn).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
//                DemoAPI.getDetail(136651365l, new RestVolleyCallback<String>() {
//                    @Override
//                    public void onSuccess(int statusCode, String data, Map<String, String> headers, boolean notModified, long networkTimeMs, String message) {
//                        showMessage(data);
//                    }
//
//                    @Override
//                    public void onFail(int statusCode, String data, Map<String, String> headers, boolean notModified, long networkTimeMs, String message) {
//                        showMessage(data);
//                    }
//                });
                DemoAPI.requestRecommendCircles(0, 20, AccountManager.instance().getUserToken(), new RestVolleyCallback<String>() {
                    @Override
                    public void onSuccess(int statusCode, String data, Map<String, String> headers, boolean notModified, long networkTimeMs, String message) {
                        showMessage(data);
                    }

                    @Override
                    public void onFail(int statusCode, String data, Map<String, String> headers, boolean notModified, long networkTimeMs, String message) {
                        showMessage(data);
                    }
                });
            }
        });

        findViewById(R.id.post_btn).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                DemoAPI.requestGuestAccount(new RestVolleyCallback<String>() {
                    @Override
                    public void onSuccess(int statusCode, String data, Map<String, String> headers, boolean notModified, long networkTimeMs, String message) {
                        showMessage(data);
                    }

                    @Override
                    public void onFail(int statusCode, String data, Map<String, String> headers, boolean notModified, long networkTimeMs, String message) {
                        showMessage(data);
                    }
                });
            }
        });

        findViewById(R.id.sync_get_btn).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                TaskScheduler.execute(new Runnable() {
                    @Override
                    public void run() {
                        TaskScheduler.execute(new Runnable() {
                            @Override
                            public void run() {
                                final RestVolleyResponse<String> result = DemoAPI.getTagSync(136651365l);
                                runOnUiThread(new Runnable() {
                                    @Override
                                    public void run() {
                                        showMessage(result.data);
                                    }
                                });
                            }
                        });
                    }
                });
            }
        });

        findViewById(R.id.sync_post_btn).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                TaskScheduler.execute(new Runnable() {
                    @Override
                    public void run() {
                        final RestVolleyResponse<String> result = DemoAPI.syncPostGuestAccount();
                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                showMessage(result.data);
                            }
                        });
                    }
                });
            }
        });

        findViewById(R.id.display_bitmap).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String url = "http://st.hujiang.com/images/bg11_s.jpg";
                RestVolleyImageLoader.instance(MainActivity.this).displayImage(url, showImage(null));
            }
        });

        findViewById(R.id.load_bitmap).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String url = "http://i2.hjfile.cn/f96/68/57/27316857.jpg";
                RestVolleyImageLoader.instance(MainActivity.this).loadImage(url, new ImageLoaderCompat.ImageListener() {
                    @Override
                    public void onResponse(ImageLoaderCompat.ImageContainer response, boolean isImmediate) {
                        showImage(response.getBitmap());
                    }

                    @Override
                    public void onErrorResponse(VolleyError error) {

                    }
                });
            }
        });

        findViewById(R.id.sync_load_bitmap).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                TaskScheduler.execute(new Runnable() {
                    @Override
                    public void run() {
                        String url = "http://i2.hjfile.cn/f96/67/87/9686787.jpg";
                        final Bitmap bitmap = RestVolleyImageLoader.instance(MainActivity.this).syncLoadImage(url);
                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                showImage(bitmap);
                            }
                        });
                    }
                });
            }
        });
    }

    private View.OnClickListener mOnClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            int id = v.getId();
            Context context = MainActivity.this;
            if (id == R.id.league_btn) {
                if (!AccountManager.instance().isLogin()) {
                    LoginActivity.start(MainActivity.this);
                } else {
                    startActivity(new Intent(MainActivity.this, LeagueActivity.class));
                }
            } else if (id == R.id.file_img_btn) {
                String path = "/mnt/sdcard/img_default.jpg";
                if (!new File(path).exists()) {
                    ToastUtils.show(MainActivity.this, "缺少磁盘文件/sdcard/img_default.jpg");
                    return;
                }
                path = "file://" + path;
                RestVolleyImageLoader.instance(context).displayImage(path, showImage(null));
            } else if (id == R.id.asset_img_btn) {
                String path = "assets://assets_default.png";
                RestVolleyImageLoader.instance(context).displayImage(path, showImage(null));
            } else if (id == R.id.content_img_btn) {
                String path = "content://media/external/images/media/27916";
                RestVolleyImageLoader.instance(context).displayImage(path, showImage(null));
            } else if (id == R.id.drawable_btn) {
                String path = "drawable://" + R.drawable.drawable_default;
                RestVolleyImageLoader.instance(context).displayImage(path, showImage(null));
            } else if (id == R.id.raw_btn) {
                String path = "drawable://" + R.raw.raw_default;
                RestVolleyImageLoader.instance(context).displayImage(path, showImage(null));
            } else if (id == R.id.download_btn) {
                //download event
//                String url = "http://app.m.hjfile.cn/android/hjwordgames_hjpc.apk";
//                String localPath = "/mnt/sdcard/hjwordgames_hjpc.apk";
                String url = "https://fbookmedia.files.wordpress.com/2015/09/360-video-in-news-feed-screenshots.zip";
                String localPath = "/mnt/sdcard/360-video-in-news-feed-screenshots.zip";
                RestVolleyDownload.download(MainActivity.this, url, localPath, new RestVolleyDownload.OnDownloadListener() {

                    @Override
                    public void onDownloadStart(String url) {

                    }

                    @Override
                    public void onDownloadSuccess(String url, File file, int httpCode, Headers headers) {
                        showMessage("success" + file.getAbsolutePath());
                    }

                    @Override
                    public void onDownloadFailure(String url, Exception e, int httpCode, Headers headers) {
                        showMessage("fail" + e.getMessage());
                    }

                    @Override
                    public void onDownloadProgress(String url, int downloadBytes, int contentLength, File file, int httpCode, Headers headers) {
                        Log.i("liuxiaoming", "progress:" + downloadBytes + "/" + contentLength);
                    }
                });
            }
        }
    };

    @Override
    public void onLogin(UserInfo userInfo) {
        updateUserProfileButtonText(AccountManager.instance().isLogin());
    }

    @Override
    public void onLogout() {
        updateUserProfileButtonText(AccountManager.instance().isLogin());
    }

    @Override
    public void onModifyAccount(UserInfo userInfo) {

    }

    private void updateUserProfileButtonText(boolean isLogin) {
        mAccountButton.setText(isLogin ? "个人中心" : "登录");
    }

    public void showMessage(CharSequence content) {
        new AlertDialog.Builder(this).setMessage(content).setNegativeButton("取消", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                dialog.dismiss();
            }
        }).setPositiveButton("确定", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                dialog.dismiss();
            }
        }).create().show();
    }

    public ImageView showImage(Bitmap bitmap) {
        LayoutInflater layoutInflater = LayoutInflater.from(this);

        View dialogView = layoutInflater.inflate(R.layout.dialog_image_layout, null);
        ImageView imageView = (ImageView)dialogView.findViewById(R.id.dialog_image);
        if (bitmap != null) {
            imageView.setImageBitmap(bitmap);
        }

        new AlertDialog.Builder(this).setView(dialogView).setNegativeButton("取消", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                dialog.dismiss();
            }
        }).setPositiveButton("确定", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                dialog.dismiss();
            }
        }).create().show();

        return imageView;
    }
}
